File: python/gbp_stack/jitter.py
"""Bounded reorder buffer for GAP audio frames."""

from __future__ import annotations

import enum
from dataclasses import dataclass, field
from typing import Dict, List, Optional


@dataclass(frozen=True)
class AudioFrame:
    """One frame held by :class:`JitterBuffer`."""

    media_source_id: int
    rtp_sequence: int
    plaintext: bytes


class JitterPushOutcome(enum.Enum):
    """Outcome of :meth:`JitterBuffer.push`."""

    ACCEPTED = "accepted"
    LATE = "late"
    EVICTED = "evicted"


@dataclass(frozen=True)
class JitterPushResult:
    """Result returned by :meth:`JitterBuffer.push`."""

    outcome: JitterPushOutcome
    evicted: Optional[AudioFrame] = None


@dataclass
class _SourceState:
    waiting: List[AudioFrame] = field(default_factory=list)
    next_seq: Optional[int] = None


class JitterBuffer:
    """Bounded reorder window keyed by ``media_source_id``.

    Holds incoming GAP frames briefly so the decoder consumes them in
    ``rtp_sequence`` order; drops anything older than the next-expected
    sequence as late.
    """

    __slots__ = ("_capacity_per_source", "_sources")

    def __init__(self, capacity_per_source: int) -> None:
        if capacity_per_source <= 0:
            raise ValueError("capacity_per_source must be > 0")
        self._capacity_per_source = capacity_per_source
        self._sources: Dict[int, _SourceState] = {}

    def push(self, frame: AudioFrame) -> JitterPushResult:
        """Insert a frame into the buffer."""
        state = self._sources.setdefault(frame.media_source_id, _SourceState())
        if state.next_seq is not None and frame.rtp_sequence < state.next_seq:
            return JitterPushResult(JitterPushOutcome.LATE)
        # Dedupe.
        if any(f.rtp_sequence == frame.rtp_sequence for f in state.waiting):
            return JitterPushResult(JitterPushOutcome.ACCEPTED)
        # Insert in sorted order.
        idx = next(
            (i for i, f in enumerate(state.waiting) if f.rtp_sequence > frame.rtp_sequence),
            len(state.waiting),
        )
        state.waiting.insert(idx, frame)
        if len(state.waiting) > self._capacity_per_source:
            evicted = state.waiting.pop(0)
            return JitterPushResult(JitterPushOutcome.EVICTED, evicted)
        return JitterPushResult(JitterPushOutcome.ACCEPTED)

    def pop_in_order(self, media_source_id: int) -> Optional[AudioFrame]:
        """Pop the next frame if its ``rtp_sequence`` is contiguous."""
        state = self._sources.get(media_source_id)
        if not state or not state.waiting:
            return None
        head = state.waiting[0]
        if state.next_seq is not None and head.rtp_sequence != state.next_seq:
            return None
        state.waiting.pop(0)
        state.next_seq = (head.rtp_sequence + 1) & 0xFFFFFFFF
        return head

    def pop_force(self, media_source_id: int) -> Optional[AudioFrame]:
        """Pop the next frame regardless of contiguity (skip gaps)."""
        state = self._sources.get(media_source_id)
        if not state or not state.waiting:
            return None
        head = state.waiting.pop(0)
        state.next_seq = (head.rtp_sequence + 1) & 0xFFFFFFFF
        return head

    def length_for(self, media_source_id: int) -> int:
        """Number of frames buffered for the given source."""
        state = self._sources.get(media_source_id)
        return len(state.waiting) if state else 0

    def clear(self) -> None:
        """Drop every queued frame."""
        self._sources.clear()

File: js/src/jitter.ts
/** Bounded reorder buffer for GAP audio frames. */

/** One frame held by {@link JitterBuffer}. */
export interface AudioFrame {
    mediaSourceId: number;
    rtpSequence: number;
    plaintext: Buffer;
}

/** Outcome of {@link JitterBuffer.push}. */
export type JitterPushOutcome = "accepted" | "late" | "evicted";

/** Result returned by {@link JitterBuffer.push}. */
export interface JitterPushResult {
    outcome: JitterPushOutcome;
    evicted?: AudioFrame;
}

interface SourceState {
    waiting: AudioFrame[];
    next: number | null;
}

/**
 * Bounded reorder window keyed by `mediaSourceId`. Holds incoming GAP
 * frames briefly so the decoder consumes them in `rtpSequence` order;
 * drops anything older than the next-expected sequence as late.
 */
export class JitterBuffer {
    private readonly capacityPerSource: number;
    private readonly sources = new Map<number, SourceState>();

    constructor(capacityPerSource: number) {
        if (capacityPerSource <= 0) throw new Error("capacityPerSource must be > 0");
        this.capacityPerSource = capacityPerSource;
    }

    /** Insert a frame into the buffer. */
    push(frame: AudioFrame): JitterPushResult {
        let state = this.sources.get(frame.mediaSourceId);
        if (!state) {
            state = { waiting: [], next: null };
            this.sources.set(frame.mediaSourceId, state);
        }
        if (state.next !== null && frame.rtpSequence < state.next) {
            return { outcome: "late" };
        }
        if (state.waiting.some(f => f.rtpSequence === frame.rtpSequence)) {
            return { outcome: "accepted" };
        }
        const idx = state.waiting.findIndex(f => f.rtpSequence > frame.rtpSequence);
        if (idx === -1) state.waiting.push(frame);
        else state.waiting.splice(idx, 0, frame);

        if (state.waiting.length > this.capacityPerSource) {
            const evicted = state.waiting.shift()!;
            return { outcome: "evicted", evicted };
        }
        return { outcome: "accepted" };
    }

    /** Pop the next frame if its `rtpSequence` is contiguous; otherwise `null`. */
    popInOrder(mediaSourceId: number): AudioFrame | null {
        const state = this.sources.get(mediaSourceId);
        if (!state || state.waiting.length === 0) return null;
        const head = state.waiting[0];
        if (state.next !== null && head.rtpSequence !== state.next) return null;
        state.waiting.shift();
        state.next = (head.rtpSequence + 1) >>> 0;
        return head;
    }

    /** Pop the next frame regardless of contiguity (skip gaps). */
    popForce(mediaSourceId: number): AudioFrame | null {
        const state = this.sources.get(mediaSourceId);
        if (!state || state.waiting.length === 0) return null;
        const head = state.waiting.shift()!;
        state.next = (head.rtpSequence + 1) >>> 0;
        return head;
    }

    /** Number of frames buffered for the given source. */
    lengthFor(mediaSourceId: number): number {
        return this.sources.get(mediaSourceId)?.waiting.length ?? 0;
    }

    /** Drops every queued frame. */
    clear(): void { this.sources.clear(); }
}
